local.go

```diff
@@ -325,20 +355,18 @@ func checkConvergence(loc *Location, iterType IterationType, stats *Stats, setti
 		}
 	}
 
-	if iterType == MajorIteration && settings.MajorIterations > 0 {
-		if stats.MajorIterations >= settings.MajorIterations {
-			return IterationLimit
-		}
-	}
 	return NotTerminated
 }
 
-// invalidate marks all fields of Location with NaNs and it is the
-// responsibility of Method to issue such EvaluationTypes that make Location
-// valid again. Invalidation is done internally as a help for implementers to
-// detect silent bugs in Methods using inconsistent Location, e.g., using
-// Gradient without a GradEvaluation request.
-func invalidate(loc *Location) {
+// evaluate evaluates the problem p at loc.X using the evaluation eval,
+// stores the answer into loc and updates stats. Unused fields of loc are set
+// to NaN. It is the responsibility of Method to assemble a valid Location
+// before requesting MajorIteration.
+func evaluate(p *Problem, loc *Location, eval Operation, stats *Stats) {
+	if !eval.IsEvaluation() {
```

Comment:
I think this should be in the minimize loop, just after the default case. Check if it's a valid evaluation, and if so evaluate.